File: ios/PSMobile/Screens/SelbsttestView.swift
import SwiftUI
import PSMShared

/// Der Selbsttest als Bildschirm: ein Knopf, eine Liste, ein Bericht.
///
/// Gedacht für das echte Gerät. Was hier grün ist, ist auf diesem iPad
/// grün — nicht im Simulator, wo es keine Speichergrenze, keine
/// Wärmegrenze und eine andere GPU gibt.
///
/// Am Ende steht eine Markdown-Datei in „Dateien → PSMobile →
/// Selbsttest". Weitergeben geht direkt über das Teilen-Blatt.
struct SelbsttestView: View {

    @StateObject private var test = Selbsttest()
    var onClose: () -> Void

    @Environment(\.psScale) private var ps

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            kopfzeile
            Divider().background(PrusaColors.divider)
            inhalt
        }
        .background(PrusaColors.background)
        .overlay(alignment: .topLeading) { PSMarke(name: "selbsttest") }
    }

    private var kopfzeile: some View {
        HStack(spacing: ps.pt(16)) {
            Button(action: onClose) {
                Text("‹  " + st("Back", "Zurück"))
                    .font(.system(size: ps.font(15)))
                    .foregroundStyle(PrusaColors.orange)
                    .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            .accessibilityIdentifier("selbsttest.zurueck")

            Text(st("Self-test", "Selbsttest"))
                .font(.system(size: ps.font(20)))
                .foregroundStyle(PrusaColors.textPrimary)
            Spacer()
            if let url = test.berichtURL {
                ShareLink(item: url) {
                    Text(st("Share report", "Bericht teilen"))
                        .font(.system(size: ps.font(13)))
                        .foregroundStyle(.white)
                        .padding(.horizontal, ps.pt(14))
                        .frame(height: ps.touch(44))
                        .background(PrusaColors.orange)
                        .clipShape(RoundedRectangle(cornerRadius: ps.pt(3)))
                        .contentShape(Rectangle())
                }
                .accessibilityIdentifier("selbsttest.teilen")
            }
        }
        .padding(.horizontal, ps.pt(16))
        .frame(height: ps.touch(56))
    }

    private var inhalt: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: ps.pt(10)) {
                Text(st("Runs everything that needs real hardware: loading, slicing, saving, reloading, painting, multi-material, and finally under load. The simulator answers none of these questions honestly.",
                        "Prüft alles, was echtes Gerät braucht: laden, slicen, sichern, wieder laden, bemalen, mehrfarbig, und zum Schluss unter Last. Der Simulator beantwortet keine dieser Fragen ehrlich."))
                    .font(.system(size: ps.font(12)))
                    .foregroundStyle(PrusaColors.textMuted)
                    .fixedSize(horizontal: false, vertical: true)

                startKnopf

                if test.laeuft, !test.aktuell.isEmpty {
                    HStack(spacing: ps.pt(8)) {
                        ProgressView()
                        Text(test.aktuell)
                            .font(.system(size: ps.font(12)))
                            .foregroundStyle(PrusaColors.textPrimary)
                    }
                }

                ForEach(test.schritte) { schritt in
                    zeile(schritt)
                }

                if !test.laeuft, !test.schritte.isEmpty {
                    abschluss
                }
            }
            .padding(ps.pt(16))
            .frame(maxWidth: ps.pt(760))
            .frame(maxWidth: .infinity)
        }
    }

    private var startKnopf: some View {
        Button {
            if test.laeuft { test.abbrechen() } else { test.starten() }
        } label: {
            Text(test.laeuft ? st("Cancel", "Abbrechen")
                             : st("Run all checks", "Alles prüfen"))
                .font(.system(size: ps.font(15), weight: .semibold))
                .foregroundStyle(.white)
                .frame(maxWidth: .infinity)
                .frame(height: ps.touch(52))
                .background(test.laeuft ? PrusaColors.panelRaised : PrusaColors.orange)
                .clipShape(RoundedRectangle(cornerRadius: ps.pt(4)))
                .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier("selbsttest.start")
    }

    private func zeile(_ schritt: Selbsttest.Schritt) -> some View {
        HStack(alignment: .top, spacing: ps.pt(10)) {
            Text(schritt.ausgang.zeichen)
                .font(.system(size: ps.font(15), weight: .bold))
                .foregroundStyle(farbe(schritt.ausgang))
                .frame(width: ps.pt(18))
            VStack(alignment: .leading, spacing: ps.pt(2)) {
                HStack {
                    Text(schritt.name)
                        .font(.system(size: ps.font(13)))
                        .foregroundStyle(PrusaColors.textPrimary)
                    Spacer()
                    Text(String(format: "%.1f s", schritt.sekunden))
                        .font(.system(size: ps.font(10)))
                        .foregroundStyle(PrusaColors.textMuted)
                }
                if !schritt.detail.isEmpty {
                    Text(schritt.detail)
                        .font(.system(size: ps.font(11)))
                        .foregroundStyle(schritt.ausgang == .fehler
                                         ? PrusaColors.danger : PrusaColors.textMuted)
                        .fixedSize(horizontal: false, vertical: true)
                }
            }
        }
        .padding(.horizontal, ps.pt(10))
        .padding(.vertical, ps.pt(8))
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(PrusaColors.panelRaised)
        .clipShape(RoundedRectangle(cornerRadius: ps.pt(4)))
        .accessibilityIdentifier("selbsttest.schritt")
    }

    private var abschluss: some View {
        VStack(alignment: .leading, spacing: ps.pt(4)) {
            Text(test.fehlerZahl == 0
                 ? st("All checks passed", "Alles bestanden")
                 : "\(test.fehlerZahl) " + st("failed", "fehlgeschlagen"))
                .font(.system(size: ps.font(15), weight: .semibold))
                .foregroundStyle(test.fehlerZahl == 0 ? PrusaColors.orange : PrusaColors.danger)
                .accessibilityIdentifier("selbsttest.ergebnis")
            if test.berichtURL != nil {
                Text(st("The report is in Files → PSMobile → Selbsttest.",
                        "Der Bericht liegt in Dateien → PSMobile → Selbsttest."))
                    .font(.system(size: ps.font(11)))
                    .foregroundStyle(PrusaColors.textMuted)
            }
        }
        .padding(.top, ps.pt(6))
    }

    private func farbe(_ ausgang: Selbsttest.Ausgang) -> Color {
        switch ausgang {
        case .ok:      return PrusaColors.orange
        case .fehler:  return PrusaColors.danger
        case .warnung: return PrusaColors.textPrimary
        case .angabe:  return PrusaColors.textMuted
        }
    }
}

File: android/app/src/main/java/de/psmobile/ui/SelbsttestView.kt
package de.psmobile.ui

import androidx.activity.compose.BackHandler
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import de.psmobile.LocalSlicerModel
import de.psmobile.diagnose.Selbsttest
import de.psmobile.ui.theme.PrusaColors
import java.util.Locale
import kotlin.concurrent.thread

/// Der Selbsttest als Bildschirm: ein Knopf, eine Liste, ein Bericht.
///
/// Gedacht für das echte Gerät. Was hier grün ist, ist auf diesem
/// Gerät grün — nicht im Emulator. Am Ende steht eine Markdown-Datei,
/// die sich direkt teilen laesst.
@Composable
fun SelbsttestView(onClose: () -> Unit) {
    val ps = LocalPsScale.current
    val context = LocalContext.current
    val model = LocalSlicerModel.current
    val schalter = LocalSlicerModel.current.startargumente
    val test = remember { Selbsttest(context, schalter) }

    val schritte by test.schritte.collectAsState()
    val aktuell by test.aktuell.collectAsState()
    val laeuft by test.laeuft.collectAsState()
    val bericht by test.bericht.collectAsState()
    val fehlerZahl = schritte.count { it.ausgang == Selbsttest.Ausgang.FEHLER }

    BackHandler(onBack = onClose)

    Box(
        Modifier
            .fillMaxSize()
            .background(PrusaColors.background),
    ) {
        Column(Modifier.fillMaxSize(), horizontalAlignment = Alignment.Start) {
            // Kopfzeile
            Row(
                Modifier
                    .fillMaxWidth()
                    .height(ps.touch(56))
                    .padding(horizontal = ps.pt(16)),
                horizontalArrangement = Arrangement.spacedBy(ps.pt(16)),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(
                    "‹  " + st("Back", "Zurück"),
                    Modifier
                        .clickable(onClick = onClose)
                        .testTag("selbsttest.zurueck"),
                    fontSize = ps.font(15), color = PrusaColors.orange,
                )
                Text(st("Self-test", "Selbsttest"), fontSize = ps.font(20), color = PrusaColors.textPrimary)
                Spacer(Modifier.weight(1f))
                val url = bericht
                if (url != null) {
                    Box(
                        Modifier
                            .height(ps.touch(44))
                            .clip(RoundedCornerShape(ps.pt(3)))
                            .background(PrusaColors.orange)
                            .clickable { teilen(context, model, url, "text/markdown") }
                            .testTag("selbsttest.teilen")
                            .padding(horizontal = ps.pt(14)),
                        contentAlignment = Alignment.Center,
                    ) {
                        Text(st("Share report", "Bericht teilen"), fontSize = ps.font(13), color = Color.White)
                    }
                }
            }
            HorizontalDivider(color = PrusaColors.divider)

            // Inhalt
            Column(
                Modifier
                    .fillMaxWidth()
                    .verticalScroll(rememberScrollState())
                    .padding(ps.pt(16)),
                verticalArrangement = Arrangement.spacedBy(ps.pt(10)),
                horizontalAlignment = Alignment.Start,
            ) {
                Text(
                    st(
                        "Runs everything that needs real hardware: loading, slicing, saving, reloading, painting, multi-material, and finally under load. The simulator answers none of these questions honestly.",
                        "Prüft alles, was echtes Gerät braucht: laden, slicen, sichern, wieder laden, bemalen, mehrfarbig, und zum Schluss unter Last. Der Simulator beantwortet keine dieser Fragen ehrlich.",
                    ),
                    fontSize = ps.font(12), color = PrusaColors.textMuted,
                )

                // Startknopf
                Box(
                    Modifier
                        .fillMaxWidth()
                        .height(ps.touch(52))
                        .clip(RoundedCornerShape(ps.pt(4)))
                        .background(if (laeuft) PrusaColors.panelRaised else PrusaColors.orange)
                        .clickable {
                            if (laeuft) {
                                test.abbrechen()
                            } else {
                                // Der Durchlauf blockiert - ein eigener
                                // Faden, der auch weiterlaeuft, wenn
                                // dieser Bildschirm zugeht.
                                thread(name = "Selbsttest") { test.durchlauf() }
                            }
                        }
                        .testTag("selbsttest.start"),
                    contentAlignment = Alignment.Center,
                ) {
                    Text(
                        if (laeuft) st("Cancel", "Abbrechen") else st("Run all checks", "Alles prüfen"),
                        fontSize = ps.font(15), fontWeight = FontWeight.SemiBold, color = Color.White,
                    )
                }

                val schrittName = aktuell
                if (laeuft && !schrittName.isNullOrEmpty()) {
                    Row(
                        horizontalArrangement = Arrangement.spacedBy(ps.pt(8)),
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        CircularProgressIndicator(Modifier.size(ps.pt(18)), color = PrusaColors.orange, strokeWidth = 2.dp)
                        Text(schrittName, fontSize = ps.font(12), color = PrusaColors.textPrimary)
                    }
                }

                schritte.forEach { schritt ->
                    Zeile(schritt)
                }

                if (!laeuft && schritte.isNotEmpty()) {
                    Abschluss(fehlerZahl, bericht != null)
                }
            }
        }
        PSMarke(name = "selbsttest")
    }
}

@Composable
private fun Zeile(schritt: Selbsttest.Schritt) {
    val ps = LocalPsScale.current
    Row(
        Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(ps.pt(4)))
            .background(PrusaColors.panelRaised)
            .testTag("selbsttest.schritt")
            .padding(horizontal = ps.pt(10), vertical = ps.pt(8)),
        horizontalArrangement = Arrangement.spacedBy(ps.pt(10)),
        verticalAlignment = Alignment.Top,
    ) {
        Text(
            schritt.ausgang.zeichen,
            Modifier.width(ps.pt(18)),
            fontSize = ps.font(15), fontWeight = FontWeight.Bold, color = farbe(schritt.ausgang),
        )
        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(ps.pt(2))) {
            Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
                Text(schritt.name, fontSize = ps.font(13), color = PrusaColors.textPrimary)
                Spacer(Modifier.weight(1f))
                Text(
                    String.format(Locale.US, "%.1f s", schritt.sekunden),
                    fontSize = ps.font(10), color = PrusaColors.textMuted,
                )
            }
            if (schritt.detail.isNotEmpty()) {
                Text(
                    schritt.detail,
                    fontSize = ps.font(11),
                    color = if (schritt.ausgang == Selbsttest.Ausgang.FEHLER) PrusaColors.danger else PrusaColors.textMuted,
                )
            }
        }
    }
}

@Composable
private fun Abschluss(fehlerZahl: Int, berichtVorhanden: Boolean) {
    val ps = LocalPsScale.current
    Column(
        Modifier.padding(top = ps.pt(6)),
        verticalArrangement = Arrangement.spacedBy(ps.pt(4)),
        horizontalAlignment = Alignment.Start,
    ) {
        Text(
            if (fehlerZahl == 0) st("All checks passed", "Alles bestanden")
            else "$fehlerZahl " + st("failed", "fehlgeschlagen"),
            Modifier.testTag("selbsttest.ergebnis"),
            fontSize = ps.font(15), fontWeight = FontWeight.SemiBold,
            color = if (fehlerZahl == 0) PrusaColors.orange else PrusaColors.danger,
        )
        if (berichtVorhanden) {
            Text(
                st(
                    "The report is in Files → PSMobile → Selbsttest.",
                    "Der Bericht liegt in Dateien → PSMobile → Selbsttest.",
                ),
                fontSize = ps.font(11), color = PrusaColors.textMuted,
            )
        }
    }
}

/// iOS kennt `warnung` ("!"), Android `UEBERSPRUNGEN` ("–") - beide
/// sind der dritte Ausgang und bekommen dieselbe Farbe.
private fun farbe(ausgang: Selbsttest.Ausgang): Color = when (ausgang) {
    Selbsttest.Ausgang.OK -> PrusaColors.orange
    Selbsttest.Ausgang.FEHLER -> PrusaColors.danger
    Selbsttest.Ausgang.UEBERSPRUNGEN -> PrusaColors.textPrimary
    Selbsttest.Ausgang.ANGABE -> PrusaColors.textMuted
}
